init/modules/EESSI/2023.06.lua

```diff
@@ -17,56 +17,120 @@ local eessi_os_type = "linux"
 setenv("EESSI_VERSION", eessi_version)
 setenv("EESSI_CVMFS_REPO", eessi_repo)
 setenv("EESSI_OS_TYPE", eessi_os_type)
+function eessiDebug(text)
+    if (mode() == "load" and os.getenv("EESSI_DEBUG_INIT")) then
+        LmodMessage(text)
+    end
+end
 function archdetect_cpu()
     local script = pathJoin(eessi_prefix, 'init', 'lmod_eessi_archdetect_wrapper.sh')
-    if not os.getenv("EESSI_ARCHDETECT_OPTIONS") then
+    -- make sure that we grab the value for architecture before the module unsets the environment variable (in unload mode)
+    local archdetect_options = os.getenv("EESSI_ARCHDETECT_OPTIONS") or (os.getenv("EESSI_ARCHDETECT_OPTIONS_OVERRIDE") or "")
+    if not os.getenv("EESSI_ARCHDETECT_OPTIONS_OVERRIDE") then
         if convertToCanonical(LmodVersion()) < convertToCanonical("8.6") then
-            LmodError("Loading this modulefile requires using Lmod version >= 8.6, but you can export EESSI_ARCHDETECT_OPTIONS to the available cpu architecture in the form of: x86_64/intel/haswell:x86_64/generic or aarch64/neoverse_v1:aarch64/generic")
+            LmodError("Loading this modulefile requires using Lmod version >= 8.6, but you can export EESSI_ARCHDETECT_OPTIONS_OVERRIDE to the available cpu architecture in the form of: x86_64/intel/haswell:x86_64/generic or aarch64/neoverse_v1:aarch64/generic")
         end
         source_sh("bash", script)
     end
-    local archdetect_options = os.getenv("EESSI_ARCHDETECT_OPTIONS") or ""
-    for archdetect_filter_cpu in string.gmatch(archdetect_options, "([^" .. ":" .. "]+)") do
-        if isDir(pathJoin(eessi_prefix, "software", eessi_os_type, archdetect_filter_cpu, "software")) then
-            -- use x86_64/amd/zen3 for now when AMD Genoa (Zen4) CPU is detected, 
-            -- since optimized software installations for Zen4 are a work-in-progress, 
-            -- see https://gitlab.com/eessi/support/-/issues/37 
-            if archdetect_filter_cpu == "x86_64/amd/zen4" then
-                archdetect_filter_cpu = "x86_64/amd/zen3"
-                if mode() == "load" then
-                    LmodMessage("Sticking to " .. archdetect_filter_cpu .. " for now, since optimized installations for AMD Genoa (Zen4) are a work in progress.")
+    -- EESSI_ARCHDETECT_OPTIONS is set by the script (_if_ it was called)
+    archdetect_options = os.getenv("EESSI_ARCHDETECT_OPTIONS") or archdetect_options
+    if archdetect_options then
+        eessiDebug("Got archdetect CPU options: " .. archdetect_options)
+        for archdetect_filter_cpu in string.gmatch(archdetect_options, "([^" .. ":" .. "]+)") do
```

Comment:
I think this for loop can use a comment on what it does. I _think_ `archdetect_options` is a colon-separated list of CPU architectures that are compatible with the host CPU (are they ordered in some way? I guess from most specific to least specific?). This loop then loops over those to check if EESSI _supports_ one of these, and selects (`return`s) the first arch for which a matching directory exists.

Reply:
That is correct!
```
ocaisa@LAPTOP-O6HF2IKC:~$ /cvmfs/software.eessi.io/versions/2023.06/init/eessi_archdetect.sh -a cpupath
x86_64/intel/skylake_avx512:x86_64/intel/haswell:x86_64/generic
```

Reply:
Done in https://github.com/EESSI/software-layer/pull/781/commits/6d64c4283c1de2fa7b9c51a1d53e78e18c7b5ae3

Reply:
I'd still add that last point though: "We loop over the list, and return the first arch for which a directory exists". (I could _guess_ what the `string.gmatch` does, but probably good to make that explicit)

Reply:
Done in https://github.com/EESSI/software-layer/pull/781/commits/7e4f901762cd7981740da8c2f8b5cc9413f0c2df